test/test_application.py

```diff
@@ -1,10 +1,70 @@
 from fastapi.testclient import TestClient
 
-from src.application import app
+from src.application import app, users_id_db
+
+client = TestClient(app)
 
 
 def test_startup():
-    with TestClient(app) as client:
-        response = client.get("/")
-        assert response.status_code == 200
-        assert response.text == "Hello, I'm good!"
+    response = client.get("/")
+    assert response.status_code == 200
+    assert response.text == "Hello, I'm good!"
+
+
+def test_post_user(mock_data_user):
+    response = client.post(
+        "/users/",
+        params=mock_data_user,
+    )
+    data = response.json()
+    assert "user_id" in data
+    assert data["email"] == mock_data_user["email"]
+    assert data["username"] == mock_data_user["username"]
+    assert data["age"] == mock_data_user["age"]
+    assert data["country"] == mock_data_user["country"]
+    assert response.status_code == 200
+
+
+def test_post_superuser(mock_data_superuser_complete):
+    response = client.post("/superusers/", params=mock_data_superuser_complete)
+    data = response.json()
+    assert "superuser_id" in data
+    assert data["user_id"] == mock_data_superuser_complete["user_id"]
+    assert data["email"] == mock_data_superuser_complete["email"]
+    assert data["username"] == mock_data_superuser_complete["username"]
+    assert data["age"] == mock_data_superuser_complete["age"]
+    assert data["country"] == mock_data_superuser_complete["country"]
+    assert data["role"] == mock_data_superuser_complete["role"]
+    assert response.status_code == 200
+
+
+def test_post_activity(mock_data_activity):
+    users_id_db.append(mock_data_activity["user_id"])
+    response = client.post("/activities/", params=mock_data_activity)
+    data = response.json()
+    assert "activity_id" in data
+    assert "time" in data
+    assert data["user_id"] == mock_data_activity["user_id"]
+    assert data["activity_type"] == mock_data_activity["activity_type"]
+    assert data["activity_details"] == mock_data_activity["activity_details"]
+    assert response.status_code == 200
+
+
+def test_filter_activities_by_user_id(mock_data_user, mock_data_activity):
+    users_id_db.append(mock_data_activity["user_id"])
+    client.post("/activities/", json=mock_data_activity)
+    response = client.get(f"/activities/?user_id={mock_data_user["user_id"]}")
+    data = response.json()
+    assert data[0]["user_id"] == mock_data_activity["user_id"]
+    assert data[0]["activity_type"] == mock_data_activity["activity_type"]
+    assert data[0]["activity_details"] == mock_data_activity["activity_details"]
+    assert response.status_code == 200
+
+
+def test_fail_filter_activities_user_id_not_found(mock_data_user, mock_data_activity):
+    client.post("/users/", json=mock_data_user)
+    client.post("/activities/", json=mock_data_activity)
+    response = client.get(f"/activities/?user_id={mock_data_user["user_id"] + 1}")
```

Comment:
as mentioned, I'd rather not call "post" AND "get" in the same unit test, as here we're testing only "get". You can define specific users in a test user database and then make sure you're "getting" the user you know exists (or doesn't, if you want to test fail) in the database. :)

Reply:
Aha! Line 65 is actually unnecessary, as user_id is already added to the user_id_db list in the previous test (and it should be fixtured for the whole session). By deleting line 65 everything runs smooth :)